File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null || head.next == null) return true;
        ListNode slow=head;
        ListNode fast=head;

        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        ListNode second=slow;
        ListNode prev=null;
        while(second!=null){
            ListNode nexttemp=second.next;
            second.next=prev;
            prev=second;
            second=nexttemp;
        }
        
            ListNode first=head;
            second=prev;
            while(second!=null){
                if(first.val!=second.val){return false;}
                    first=first.next;
                    second=second.next;
                
            }
            return true;
    }
}

File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:

        slow=head
        fast=head
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next


        curr=slow
        prev=None

        while curr:
            next=curr.next
            curr.next=prev
            prev=curr
            curr=next
        
        first=head
        sec=prev

        while sec:
            if first.val!=sec.val:
                return False
            first=first.next
            sec=sec.next
        
        return True
        
        